File: src/solution_java/medium/_318MaximumProductOfWordLengths.java
package solution_java.medium;

public class _318MaximumProductOfWordLengths {
    public class Solution {
        public int maxProduct(String[] words) {
            int size = words.length;
            int[] masks = new int[size];
            for (int j = 0; j < size; j++) {
                String word = words[j];
                int mask = 0;
                for (int i = 0; i < word.length(); i++) {
                    char ch = word.charAt(i);
                    mask = mask | (1 << (ch - 'a'));
                }
                masks[j] = mask;
            }
            int result = 0;
            for (int i = 0; i < size; i++) {
                for (int j = 0; j < i; j++) {
                    if ((masks[j] & masks[i]) == 0) {
                        result = Math.max(result, words[j].length() * words[i].length());
                    }
                }
            }
            return result;
        }
    }
}

File: src/medium/_318MaximumProductOfWordLengths.kt
package medium

class _318MaximumProductOfWordLengths {
    class Solution {
        fun maxProduct(words: Array<String>): Int {
            val n = words.size
            var idx = 0
            val masks = IntArray(n)
            for (w in words) {
                var t = 0
                for (i in w.indices) {
                    val u = w[i] - 'a'
                    t = t or (1 shl u)
                }
                masks[idx] = t
                idx += 1
            }
            var ans = 0
            for (i in 0 until n) {
                for (j in 0 until i) {
                    if ((masks[i] and masks[j]) == 0) {
                        ans = Math.max(ans, words[i].length * words[j].length)
                    }
                }
            }
            return ans
        }
    }
}